libtcmu.c

```diff
@@ -383,18 +383,192 @@ static void tcmu_cdb_debug_info(const struct tcmulib_cmd *cmd)
 		free(buf);
 }
 
-static void *tcmu_cmdproc_thread(void *arg)
+static void _cleanup_mutex_lock(void *arg)
+{
+	pthread_mutex_unlock(arg);
+}
+
+static void *aio_work_queue(void *arg)
 {
 	struct tcmu_device *dev = arg;
+	struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+	while (1) {
+		int ret;
+		struct tcmu_aio_entry *aio_entry;
+
+		pthread_cleanup_push(_cleanup_mutex_lock, &aio_wq->aio_lock);
+		pthread_mutex_lock(&aio_wq->aio_lock);
+
+		while (list_empty(&aio_wq->aio_queue)) {
+			pthread_cond_wait(&aio_wq->aio_cond,
+					  &aio_wq->aio_lock);
+		}
+
+		aio_entry = list_first_entry(&aio_wq->aio_queue,
+					     struct tcmu_aio_entry, entry);
+		tcmulib_track_aio_request(dev, &aio_entry->entry);
```

Comment:
What's the purpose of tracking the "io" request? Is it just to try to batch wakeups to the kernel? Couldn't that be done w/ a spinlock around an integer? Is this going to be re-used by true aio devices to batch kernel wakeups?